ethpm_cli/scraper.py

```diff
@@ -0,0 +1,215 @@
+    """
+    Scrapes VersionRelease event data starting from start_block.
+    """
+    chain_data_path = ethpm_dir / "chain_data.json"
+    try:
+        active_block = import_ethpm_dir(ethpm_dir, w3, start_block)
+        scraped_manifests = scrape_block_for_manifests(w3, active_block)
+
+        update_chain_data(chain_data_path, active_block, scraped_manifests)
+        write_ipfs_uris_to_disk(ethpm_dir, scraped_manifests)
+
+    except BlockNotFoundError:
+        logger.info("No more blocks available.")
+        return None
+
+    except BlockAlreadyScrapedError:
+        active_block = start_block + 1
+
+    scrape(w3, ethpm_dir, active_block)
+
+
+def import_ethpm_dir(ethpm_dir: Path, w3: Web3, start_block: int) -> int:
+    """
+    Returns the latest processed block number found in ethpm_dir.
+    If no ethpm_dir found, creates an ethpm_dir and returns a 0.
+    """
+    if ethpm_dir.is_dir():
+        chain_data_path = ethpm_dir / "chain_data.json"
+        validate_chain_data_store(chain_data_path, w3)
+
+        if start_block >= w3.eth.blockNumber:
+            raise BlockNotFoundError(
+                f"Block number: {start_block} not available on provided web3 "
+                f"instance with latest block number of {w3.eth.blockNumber}."
+            )
+
+        all_scraped_blocks = get_scraped_blocks(chain_data_path)
+        if start_block in all_scraped_blocks:
+            raise BlockAlreadyScrapedError(
+                f"Skipping block #{start_block}. Already processed."
+            )
+        return start_block
+    else:
+        ethpm_dir.mkdir()
+        # make base dir for all ipfs assets
+        (ethpm_dir / "Qm").mkdir()
+        chain_data_path = ethpm_dir / "chain_data.json"
+        chain_data_path.touch()
+        init_json = {
+            "chain_id": w3.eth.chainId,
+            "registry_addresses": [],
+            "scraped_blocks": [{"min": "0", "max": "0"}],
+        }
+        chain_data_path.write_text(f"{json.dumps(init_json, indent=4)}\n")
+        return 1
+
+
+def update_chain_data(
+    chain_data_path: Path, block_number: int, manifests: Dict[Address, Dict[str, str]]
+) -> None:
+    chain_data = json.loads(chain_data_path.read_text())
+
+    new_addrs = list(manifests.keys())
+    updated_addrs = list(sorted(set(chain_data["registry_addresses"] + new_addrs)))
+
+    all_scraped_blocks = get_scraped_blocks(chain_data_path)
+    updated_blocks = blocks_to_ranges(set(all_scraped_blocks + [block_number]))
+
+    chain_data_updated_blocks = assoc(chain_data, "scraped_blocks", updated_blocks)
+    chain_data_updated_addrs = assoc(
+        chain_data_updated_blocks, "registry_addresses", updated_addrs
+    )
+    chain_data_path.write_text(f"{json.dumps(chain_data_updated_addrs, indent=4)}\n")
+
+
+@to_list
+def blocks_to_ranges(blocks_list: List[int]) -> Iterable[Dict[str, str]]:
+    """
+    Takes a list of block numbers, and returns them grouped into min/max ranges.
```

Comment:
I *think* I understand what this says but I'm not sure.  Can you maybe provide a small code example in this docstring so it is illustrative?

Is this what it does?

https://github.com/ethereum/trinity/pull/569/files#diff-3c5a1b448f91c1258a0e43b661b3af54R16

Reply:
yup! more or less, just formatted slightly differently